### ISSUE TITLE
Wayland window issues on Plasma

### ISSUE THREAD

**Issue by aufkrawall** [mpv-player/mpv#7043]
### mpv version and platform

git-master (+ https://github.com/mpv-player/mpv/pull/7037)

Few different paper cuts:

### Reproduction steps

1st case: Open a video with bigger resolution than screen.

2nd case: Use `--autofit-larger=94%`.

3rd case: Make the the mouse cursor visible on top of mpv window, e.g. when moving or placing it on top of pseudo GUI bar.

4th case: Double click into the video.

### Expected behavior

1st case: Like on Xorg, video should be downscaled to fit the screen's resolution.

2nd case: It works on Xorg and Gnome Wayland. It should work the same on Plasma Wayland (if it's not a Plasma bug).

3rd case: It should follow Plasma's cursor theming.

4th case: It should go into fullscreen and stay there.

### Actual behavior

1st case: Video dimensions are clipped instead. It still is clipped even when going into fullscreen afterwards. To make it properly downscale, I need to maximize the window first. Then it also looks correct after switching into fullscreen.

2nd case: It causes mpv's window to not appear at all. Instead, there are lots of errors:
> [vo/gpu/opengl] after creating texture: OpenGL error INVALID_OPERATION. 
> [vo/gpu/opengl] after rendering: OpenGL error INVALID_FRAMEBUFFER_OPERATION.

3rd case: The cursor ignores Plasma's theming and instead is just black.

4th case: It instead seems to instantly leave fullscreen again and window decorations break.
A single click on the fullscreen button of the pseudo GUI works, as does pressing f.

Sometimes also Plasma's task bar doesn't get hidden in fullscreen, but that seems to be a bug of Plasma where it erroneously places the task bar on top of mpv's and other application's fullscreen window.

**Comment by duracellcell** [comment]
https://phabricator.kde.org/T10867
https://phabricator.kde.org/T4455
https://phabricator.kde.org/project/board/99/

**Comment by Dudemanguy911** [comment]
I actually managed to get it to work on this computer. The last time I tried, I just got mysterious errors. Well anyways the offending PRs are #6978 and #7012. It working before was actually just a fluke. mpv didn't actually always create the `wl_output` on init (it happened later) which led to issues like broken windows on multi monitor setups with certain options. I think what's happening here (like what @duracellcell's links suggest) is that Plasma is not handling geometry properly of windows larger than the display. That shouldn't be our job. That is up to the compositor to do correctly. Going back to not creating the `wl_output` is not an option since it would reopen some other issues and also mpv is hilariously broken on weston before those PRs in this case.

Sidenote: I actually can't reproduce the 3rd issue on the current master or my latest PR. Maybe it's because I personally use breeze-snow which is a Plasma theme? It's actually broken before the two wl_output PRs amusingly enough.

P.S. I don't really consider 1 an issue. I think it's just a choice of the compositor whether or not to do this. Weston doesn't resize the window automatically but I don't see anything wrong with that. Sway, of course, does (tiling and all that).

Anyway TL;DR, Plasma needs to handle window geometry correctly.

**Comment by aufkrawall** [comment]
I checked out Sway and it indeed just works.

I read that Wayland windows are required to select a cursor theme by themselves. Probably not realistic to expect, outside of GTK and Qt?

**Comment by Dudemanguy911** [comment]
It should read the cursor theme thanks to #6202. It did for me on plasma anyway. Maybe that environment variable isn't being set for you for some reason?

**Comment by aufkrawall** [comment]
Hm, the commit name mentions only cursor size, not theme.
Both theme and size are correctly exported.

Setting it via XDG spec helped:
https://wiki.archlinux.org/index.php/Cursor_themes#XDG_specification

**Comment by Dudemanguy** [comment]
Plasma 5.17.0 has a lot of wayland-related [changes](https://kde.org/announcements/plasma-5.16.5-5.17.0-changelog.php) and some of them seem to be about client geometry. If we're lucky, some of the above is fixed now.

**Comment by glaubersm** [comment]
4th case also occurs on Gnome Wayland session and it's a known mpv bug.
It has nothing to do with Plasma.

**Comment by Dudemanguy** [comment]
>4th case also occurs on Gnome Wayland session and it's a known mpv bug.

Works just fine for me. All of the stuff in the OP only occurs on Plasma.

**Comment by aufkrawall** [comment]
I've noticed that also on Weston the window isn't downscaled by default and instead is cut with bigger video resolution than screen resolution. Though it looks correct then when switching to fullscreen, without the need to maximize the window before (which doesn't seem to be possible on Weston?).

**Comment by glaubersm** [comment]
I have reproduced 4th casa on Gnome 3.34.1 Wayland with mpv 0.29.1 right now.

**Comment by Dudemanguy** [comment]
Please use the master branch. 0.29.1 is quite old and there have been quite a few wayland changes since then.

**Comment by aufkrawall** [comment]
It seems to me that there is a conflict of the fullscreen switch and the mouse cursor changing into drag & drop mode when double clicking on at least Gnome.
But I suppose this is managed by the compositor and not mpv's fault.

**Comment by glaubersm** [comment]
I have just installed mpv-git on Arch Linux runing Gnome 3.34.1 Wayland, I can confirm that cursor changes to "drag-and-drop" mode immediatelly after the first click when I switch to full screen mode
(or try to do it) with double click/tap using a touchpad.

Here mpv-git goes to full screen mode and immediately leaves it if I do double-tap with the touchpad.
If I do double-click using the left button of the touchpad, mpv goes to full screen mode and stay there as expected.

**Comment by Dudemanguy** [comment]
>Here mpv-git goes to full screen mode and immediately leaves it if I do double-tap with the touchpad.

If I'm interpreting you correctly, this sounds like a potential problem. Please open a new issue for it since this one is meant for Plasma and we should avoid cluttering it up.

**Comment by glaubersm** [comment]
In fact double-tap is ignored.
mpv-git goes to full screen mode and immediately leaves it if I do triple-tap, not double-tap.
It's a known bug
https://github.com/mpv-player/mpv/issues/6633

**Comment by aufkrawall** [comment]
@Dudemanguy When playing a video with resolution bigger than of the screen and then make the window floating via Mod1 + Shift + Space, mpv switches into fullscreen instead on Sway. This doesn't happen with video resolutions that are lower than of the screen.

**Comment by Dudemanguy** [comment]
It's not actually fullscreen. It's just a window that's exactly the same size as the monitor's resolution. You can still drag it around and stuff.

**Comment by aufkrawall** [comment]
Ok. However, when decreasing its size, the window title bar doesn't cover the window's full length.
A bug of Sway?

**Comment by Dudemanguy** [comment]
Yeah that's an old issue and rendering title bars isn't our problem. IIRC, rooston's title bars (I think it had titlebars anyway) didn't have this issue with resizing.

**Comment by Dudemanguy** [comment]
I revisited this one given some of the latest geometry/resizing changes. So this is with Plasma 5.17.5 + the latest mpv from git.

1. As discussed above, not a bug but yeah it still occurs.
2. Works for me.
3. Works for me (if not, set the mouse cursor to the XDG spec but I think Plasma probably does this anyway).
4. Not a wayland bug but the double click/tap thing as discussed earlier. That has its own bug report.

Anyways, closing this one now since 2 works and the rest are different things.

**Comment by aufkrawall** [comment]
Plasma guys have fixed all issues, kwinft also supports presentation time protocol.

It's just the cursor that still deviates, while XCURSOR_THEME=Vanilla-DMZ env is set.

**Comment by ApatheticSenshi** [comment]
SOLUTION - Right click the title bar of the application. One of the options of the context menu is “Keep above others”
https://imgur.com/a/tD5XDxP
